File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/execution/CqlNulLogicalOperatorsTest.java
package org.opencds.cqf.cql.engine.execution;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.math.BigDecimal;
import java.util.Collections;
import org.junit.jupiter.api.Test;
import org.opencds.cqf.cql.engine.elm.executing.EquivalentEvaluator;
import org.opencds.cqf.cql.engine.runtime.DateTime;
import org.opencds.cqf.cql.engine.runtime.Time;

class CqlNulLogicalOperatorsTest extends CqlTestBase {

    @Test
    void all_null_logical_operators() {
        final BigDecimal bigDecimalZoneOffset = getBigDecimalZoneOffset();
        var results = engine.evaluate(toElmIdentifier("CqlNullologicalOperatorsTest"));
        var value = results.forExpression("CoalesceANull").value();
        assertThat(value, is("a"));

        value = results.forExpression("CoalesceNullA").value();
        assertThat(value, is("a"));

        value = results.forExpression("CoalesceEmptyList").value();
        assertThat(value, is(nullValue()));

        value = results.forExpression("CoalesceListFirstA").value();
        assertThat(value, is("a"));

        value = results.forExpression("CoalesceListLastA").value();
        assertThat(value, is("a"));

        value = results.forExpression("CoalesceFirstList").value();
        assertThat(value, is(Collections.singletonList("a")));

        value = results.forExpression("CoalesceLastList").value();
        assertThat(value, is(Collections.singletonList("a")));

        value = results.forExpression("DateTimeCoalesce").value();
        assertTrue(EquivalentEvaluator.equivalent(value, new DateTime(bigDecimalZoneOffset, 2012, 5, 18)));

        value = results.forExpression("DateTimeListCoalesce").value();
        assertTrue(EquivalentEvaluator.equivalent(value, new DateTime(bigDecimalZoneOffset, 2012, 5, 18)));

        value = results.forExpression("TimeCoalesce").value();
        assertTrue(EquivalentEvaluator.equivalent(value, new Time(5, 15, 33, 556)));

        value = results.forExpression("TimeListCoalesce").value();
        assertTrue(EquivalentEvaluator.equivalent(value, new Time(5, 15, 33, 556)));

        value = results.forExpression("IsNullTrue").value();
        assertThat(value, is(true));

        value = results.forExpression("IsNullFalseEmptyString").value();
        assertThat(value, is(false));

        value = results.forExpression("IsNullAlsoFalseAbcString").value();
        assertThat(value, is(false));

        value = results.forExpression("IsNullAlsoFalseNumber1").value();
        assertThat(value, is(false));

        value = results.forExpression("IsNullAlsoFalseNumberZero").value();
        assertThat(value, is(false));

        value = results.forExpression("IsFalseFalse").value();
        assertThat(value, is(true));

        value = results.forExpression("IsFalseTrue").value();
        assertThat(value, is(false));

        value = results.forExpression("IsFalseNull").value();
        assertThat(value, is(false));

        value = results.forExpression("IsTrueTrue").value();
        assertThat(value, is(true));

        value = results.forExpression("IsTrueFalse").value();
        assertThat(value, is(false));

        value = results.forExpression("IsTrueNull").value();
        assertThat(value, is(false));
    }
}

File: Src/java/engine/src/test/kotlin/org/opencds/cqf/cql/engine/execution/CqlNulLogicalOperatorsTest.kt
package org.opencds.cqf.cql.engine.execution

import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.opencds.cqf.cql.engine.elm.executing.EquivalentEvaluator
import org.opencds.cqf.cql.engine.runtime.DateTime
import org.opencds.cqf.cql.engine.runtime.Time

internal class CqlNulLogicalOperatorsTest : CqlTestBase() {
    @Test
    fun all_null_logical_operators() {
        val bigDecimalZoneOffset = bigDecimalZoneOffset
        val results = engine.evaluate(toElmIdentifier("CqlNullologicalOperatorsTest"))
        var value = results.forExpression("CoalesceANull")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`("a"))

        value = results.forExpression("CoalesceNullA")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`("a"))

        value = results.forExpression("CoalesceEmptyList")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(Matchers.nullValue()))

        value = results.forExpression("CoalesceListFirstA")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`("a"))

        value = results.forExpression("CoalesceListLastA")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`("a"))

        value = results.forExpression("CoalesceFirstList")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(mutableListOf<String?>("a")))

        value = results.forExpression("CoalesceLastList")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(mutableListOf<String?>("a")))

        value = results.forExpression("DateTimeCoalesce")!!.value()
        Assertions.assertTrue(
            EquivalentEvaluator.equivalent(value, DateTime(bigDecimalZoneOffset, 2012, 5, 18)) ==
                true
        )

        value = results.forExpression("DateTimeListCoalesce")!!.value()
        Assertions.assertTrue(
            EquivalentEvaluator.equivalent(value, DateTime(bigDecimalZoneOffset, 2012, 5, 18)) ==
                true
        )

        value = results.forExpression("TimeCoalesce")!!.value()
        Assertions.assertTrue(EquivalentEvaluator.equivalent(value, Time(5, 15, 33, 556)) == true)

        value = results.forExpression("TimeListCoalesce")!!.value()
        Assertions.assertTrue(EquivalentEvaluator.equivalent(value, Time(5, 15, 33, 556)) == true)

        value = results.forExpression("IsNullTrue")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(true))

        value = results.forExpression("IsNullFalseEmptyString")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(false))

        value = results.forExpression("IsNullAlsoFalseAbcString")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(false))

        value = results.forExpression("IsNullAlsoFalseNumber1")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(false))

        value = results.forExpression("IsNullAlsoFalseNumberZero")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(false))

        value = results.forExpression("IsFalseFalse")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(true))

        value = results.forExpression("IsFalseTrue")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(false))

        value = results.forExpression("IsFalseNull")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(false))

        value = results.forExpression("IsTrueTrue")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(true))

        value = results.forExpression("IsTrueFalse")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(false))

        value = results.forExpression("IsTrueNull")!!.value()
        MatcherAssert.assertThat(value, Matchers.`is`(false))
    }
}
